### ISSUE TITLE
Arduino: binaries installed through Boards Manager fail to run

### ISSUE THREAD

**Issue by tg-x** [NixOS/nixpkgs#73965]
**Describe the bug**
Binaries installed through the Boards Manager in Arduino IDE
fail to run due to missing shared libraries.

A solution to this is to `patchelf` them manually after installation.

**To Reproduce**
1. Tools > Boards > Boards Manager
2. install Arduino SAM Boards
3. Verify/Compile

**Expected behavior**
Code is compiled using the installed binaries

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.80, NixOS, 19.03.173660.3f92c2124a5 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.03.173660.3f92c2124a5, nixos-unstable-20.03pre201791.c1966522d7d"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by veprbl** [comment]
@tg-x Which package provides "Arduino IDE"?

**Comment by tg-x** [comment]
`arduino-1.8.5`

**Comment by veprbl** [comment]
I see that we have an extra commit on master that is not on release-19.09: https://github.com/NixOS/nixpkgs/commit/4c85abae94b0a55ff76387c1f1793370a170dda8#diff-556d00d6117d82c7a9b1845340d48283
@tg-x Could you try arduino-1.8.9 from master?

cc @antono @auntie @robberer @bjornfor @bergey

**Comment by tg-x** [comment]
same issue with `arduino-1.8.9` from master

**Comment by veprbl** [comment]
@tg-x For documentation purposes, could you, please, post the error messages and manual patchelf commands that fix it for you.

**Comment by tg-x** [comment]
for the `Arduino AVR Boards` package it looks like this:

```
ldd ~/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino5/bin/avr-gcc                                                                                                                                          
        linux-vdso.so.1 (0x00007ffe11fb4000)                                                                                                                                                                                      
        libm.so.6 => /nix/store/h0p0h3rh1q4i2yavzm3yqi716s9yaj2f-glibc-2.27/lib/libm.so.6 (0x00007f145d6f4000)                                                                                                                    
        libc.so.6 => /nix/store/h0p0h3rh1q4i2yavzm3yqi716s9yaj2f-glibc-2.27/lib/libc.so.6 (0x00007f145d53e000)                                                                                                                    
        /lib64/ld-linux-x86-64.so.2 => /nix/store/h0p0h3rh1q4i2yavzm3yqi716s9yaj2f-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007f145d88c000)
```

then running `patchelf --set-interpreter` for all the binaries:

```
for bin in ~/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino5/avr/bin/* ~/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino5/avr/bin/*; do
  ld=$(ldd $bin | awk '/ld-linux/{print $3}')
  if [ "$ld" -ne "" ]; then patchelf --set-interpreter "$ld" "$bin"; fi
done
```

**Comment by artemist** [comment]
I have a simple package that puts arduino in an FHSEnv. It could probably be adapted into something in-tree eventually
```nix
with import <nixpkgs> { };
let
  unwrappedArduino = pkgs.arduino;
  in buildFHSUserEnv {
    name = "arduino";
    passthru = { inherit (unwrappedArduino) pname version meta; };
    targetPkgs = pkgs: with pkgs; [
      unwrappedArduino
      zlib
      (python3.withPackages (p: with p; [ pyserial ]))
    ];
    extraInstallCommands = ''
      mkdir -p $out/share/applications
      cp ${unwrappedArduino}/share/applications/arduino.desktop $out/share/applications/arduino.desktop
      substituteInPlace $out/share/applications/arduino.desktop \
          --replace ${unwrappedArduino}/bin/arduino $out/bin/arduino
    '';
    runScript = "arduino";
  }
```

**Comment by Stunkymonkey** [comment]
we could use `buildFHSUserEnv` with the same packages from [`nix-env`](https://github.com/nix-community/nix-environments/blob/master/envs/arduino/shell.nix)
I expect the `buildFHSUserEnv` will be done equally to [`platformio`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/embedded/platformio/chrootenv.nix).
Maybe someone with more skill than me could help here.

**Comment by Stunkymonkey** [comment]
if someone can confirm that my PR #144772 fixes this issue, we maybe could get this merged into "21.11"
